dependency-convergence/src/main/java/ArtifactData.java

```diff
@@ -0,0 +1,230 @@
+      if (lastUpdated != null && !lastUpdated.isEmpty()) {
+        Date date = dateFormat.parse(lastUpdated);
+        lastUpdated = outputFormat.format(date);
+      } else {
+        lastUpdated = "";
+      }
+      return new LatestMetadata(latestVersion, lastUpdated);
+    } catch (XmlPullParserException | ParseException | IOException ignored) {
+      return new LatestMetadata("", "");
+    }
+  }
+
+  /**
+   * Grabs the Github URL from the POM's scm section. If the URL is not found, then this checks the
+   * associated parent POM (given that the current POM is not already the parent of our artifact)
+   *
+   * @param pomUrl the Maven URL of the artifact's POM file
+   * @return artifact's Github URL
+   */
+  private static String getScmGithubUrl(String pomUrl) {
+    try {
+      File pomFile = File.createTempFile("pomFile", ".xml");
+      pomFile.deleteOnExit();
+      URL url = new URL(pomUrl);
+      FileUtils.copyURLToFile(url, pomFile);
+      MavenXpp3Reader read = new MavenXpp3Reader();
+      Model model = read.read(new FileInputStream(pomFile));
+      if (model == null) {
+        return null;
+      }
+      return model.getUrl();
+    } catch (XmlPullParserException | IOException ignored) {
+      return null;
+    }
+  }
+
+  private static String generatePomFileUrl(Artifact artifact) {
+    String groupPath = artifact.getGroupId().replace('.', '/');
+    return basePath + "/" + groupPath
+        + "/" + artifact.getArtifactId()
+        + "/" + artifact.getVersion()
+        + "/" + artifact.getArtifactId() + "-" + artifact.getVersion() + ".pom";
+  }
+
+  private static String generateMetadataUrl(Artifact artifact) {
+    String groupPath = artifact.getGroupId().replace('.', '/');
+    return basePath + "/" + groupPath
+        + "/" + artifact.getArtifactId()
+        + "/maven-metadata.xml";
+  }
+
+  private static class LatestMetadata {
+
+    String latestVersion, lastUpdated;
+
+    LatestMetadata(String latestVersion, String lastUpdated) {
+      this.latestVersion = latestVersion;
+      this.lastUpdated = lastUpdated;
+    }
+  }
+
+  private static String getSharedDependenciesVersionFromUrl(String pomUrl) {
+    if (pomUrl == null) {
+      return null;
+    }
+    try {
+      File pomFile = File.createTempFile("pomFile", ".xml");
+      pomFile.deleteOnExit();
+      URL url = new URL(pomUrl);
+      FileUtils.copyURLToFile(url, pomFile);
+      MavenXpp3Reader read = new MavenXpp3Reader();
+      Model model = read.read(new FileInputStream(pomFile));
+      if (model.getDependencyManagement() == null) {
+        return null;
+      }
+      for (Dependency dep : model.getDependencyManagement()
+          .getDependencies()) {
+        if ("com.google.cloud".equals(dep.getGroupId()) && "google-cloud-shared-dependencies"
+            .equals(dep.getArtifactId())) {
+          return dep.getVersion();
+        }
+      }
+      return "";
```

Comment:
why empty string here and null below?

Reply:
I will add this to the documentation - an empty string is given for an empty version for google-cloud-shared-dependencies (this is consistent with the way it's managed within the dashboard), whereas a null value is given when experiencing an error in finding the POM file to begin with